From: Steven Whitehouse <swhiteho@redhat.com>
Date: Wed, 25 Jun 2008 22:20:08 +0100
Subject: [gfs2] cannot use fifo nodes
Message-id: 1214428808.3657.3.camel@localhost.localdomain
O-Subject: [RHEL 5.3] bz #450276 GFS2: cannot use fifo nodes (named pipes)
Bugzilla: 450276
RH-Acked-by: Bob Peterson <rpeterso@redhat.com>

Hi,

Here are the two upstream patches merged into one and ported to RHEL
(only cosmetic changes). This has been upstream for some time and is a
known fix for this particular bug (bz #450276),

Steve.

-----------------------------------------------------------------------------------

diff --git a/fs/gfs2/inode.c b/fs/gfs2/inode.c
index fe50e63..439fc52 100644
--- a/fs/gfs2/inode.c
+++ b/fs/gfs2/inode.c
@@ -151,7 +151,8 @@ void gfs2_set_iop(struct inode *inode)
 	} else if (S_ISLNK(mode)) {
 		inode->i_op = &gfs2_symlink_iops;
 	} else {
-		inode->i_op = &gfs2_dev_iops;
+		inode->i_op = &gfs2_file_iops;
+		init_special_inode(inode, inode->i_mode, inode->i_rdev);
 	}
 
 	unlock_new_inode(inode);
diff --git a/fs/gfs2/ops_inode.c b/fs/gfs2/ops_inode.c
index 6f82ce1..a334369 100644
--- a/fs/gfs2/ops_inode.c
+++ b/fs/gfs2/ops_inode.c
@@ -1148,16 +1148,6 @@ const struct inode_operations gfs2_file_iops = {
 	.removexattr = gfs2_removexattr,
 };
 
-const struct inode_operations gfs2_dev_iops = {
-	.permission = gfs2_permission,
-	.setattr = gfs2_setattr,
-	.getattr = gfs2_getattr,
-	.setxattr = gfs2_setxattr,
-	.getxattr = gfs2_getxattr,
-	.listxattr = gfs2_listxattr,
-	.removexattr = gfs2_removexattr,
-};
-
 const struct inode_operations gfs2_dir_iops = {
 	.create = gfs2_create,
 	.lookup = gfs2_lookup,
diff --git a/fs/gfs2/ops_inode.h b/fs/gfs2/ops_inode.h
index 34f0caa..c2bcd6b 100644
--- a/fs/gfs2/ops_inode.h
+++ b/fs/gfs2/ops_inode.h
@@ -15,6 +15,5 @@
 extern const struct inode_operations gfs2_file_iops;
 extern const struct inode_operations gfs2_dir_iops;
 extern const struct inode_operations gfs2_symlink_iops;
-extern const struct inode_operations gfs2_dev_iops;
 
 #endif /* __OPS_INODE_DOT_H__ */
